File: Swift/Utils/Utils/ZipUtils.m
#import "ZipUtils.h"

@implementation ZipUtils

-(void) unzip {
    [NSThread detachNewThreadSelector:@selector(doUncompress) toTarget:self withObject:nil];
}

-(void) doUncompress {
    BOOL b = NO;
    if ([self.delegate respondsToSelector:@selector(zipWillUnzip)]) {
        b = [self.delegate zipWillUnzip];
    }
    if (b) {
        ZipArchive * za = [[ZipArchive alloc] init];
        NSNumber * ret = [NSNumber numberWithBool:NO];
        if ([za UnzipOpenFile: self.archiveFile]) {
            BOOL succ = [za UnzipFileTo:self.extractPath overWrite:YES];
            [za UnzipCloseFile];
            ret = [NSNumber numberWithBool:succ];
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            [self callback: ret];
        });
    }
}

-(void) callback: (NSNumber *)ret {
    if ([self.delegate respondsToSelector:@selector(ziputils:unzipCompleted:)]) {
        [self.delegate ziputils:self unzipCompleted:[ret boolValue]];
    }
}


@end

File: Swift/ZipUtils.swift
import UIKit

@objc
protocol ZipUtilsDelegate: NSObjectProtocol {
    optional func zipWillUnzip() -> Bool
    optional func ziputils(ziputils: ZipUtils, unzipCompleted succ: Bool)

}

class ZipUtils: NSObject {
    var delegate: ZipUtilsDelegate?
    var archiveFile: String?
    var extractPath: String?
    
    func unzip() {
        NSThread.detachNewThreadSelector("doUncompress", toTarget: self, withObject: nil)
    }
    
    func doUncompress() {
        var b = false
        if (self.delegate?.zipWillUnzip != nil) {
            b = self.delegate!.zipWillUnzip!()
        }
        if (b) {
            var za = ZipArchive()
            var ret = NSNumber(bool: false)
            if (za.UnzipOpenFile(self.archiveFile!)) {
                var succ = za.UnzipFileTo(self.extractPath, overWrite:true)
                za.UnzipCloseFile()
                ret = NSNumber(bool: succ)
            }
            dispatch_async(dispatch_get_main_queue(), {
                self.callback(ret)
            })
        }
    }
    
    func callback(ret: NSNumber) {
        self.delegate?.ziputils?(self, unzipCompleted: ret.boolValue)
    }

}
